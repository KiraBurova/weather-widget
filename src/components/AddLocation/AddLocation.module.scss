@import '../../styles/variables';

.formLabel {
  font-size: 0.9rem;
  font-weight: bold;
}

.form {
  display: flex;
  justify-content: stretch;
  align-items: stretch;
}

.addButton {
  font-size: 2rem;
  font-weight: bold;
  color: $main-dark;
  background: none;
  border: none;
  padding: 0;
  cursor: pointer;
  transition: color 0.2s ease-in-out;

  &:hover {
    color: lighten($main-dark, 20%);
  }

  // &:disabled {
  //   background-color: $disabled;
  //   cursor: initial;
  // }
}

.locationInput {
  flex: 1;
  font-size: 1rem;
  padding: 0.5em;
  border-radius: $border-radius;
  border: 1px solid $main-dark;
}
