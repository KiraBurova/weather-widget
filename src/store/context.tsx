import { useLocalObservable } from 'mobx-react-lite';
import React from 'react';

import { createStore, TStore } from './store';

const StoreContext = React.createContext<TStore | null>(null);

type StoreProviderProps = {
  children: React.ReactNode;
};

export const StoreProvider = ({ children }: StoreProviderProps) => {
  const store = useLocalObservable(createStore);
  return <StoreContext.Provider value={store}>{children}</StoreContext.Provider>;
};

export const useStore = () => {
  const store = React.useContext(StoreContext);
  if (!store) {
    throw new Error('useStore must be used within a StoreProvider.');
  }
  return store;
};
