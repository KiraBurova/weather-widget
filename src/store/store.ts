import { getWeatherReportByCityName, getWeatherReportByLocation, getWeatherIconByIconName } from '../api';

import { TLocation, WeatherReportType } from '../types';

export function createStore() {
  return {
    location: '',
    locations: [] as TLocation[],
    weatherReports: [] as WeatherReportType[],
    fetchWeatherReportByLocation(latitude: number, longitude: number) {
      getWeatherReportByLocation(latitude, longitude)
        .then((weatherData: any) => {
          console.log(weatherData);
          const w = this.transformLocationData(weatherData.data);
          this.setWeatherReports(w);
        })
        .catch((error: Error) => console.log(error));
    },
    fetchWeatherReportByCityName(location: TLocation) {
      getWeatherReportByCityName(location.name)
        .then((weatherData) => {
          this.setLocations(location);
          const w = this.transformLocationData(weatherData.data);
          this.setWeatherReports(w);
        })
        .catch((error: Error) => console.log(error));
    },
    setLocations(location: TLocation) {
      this.locations.push(location);
    },
    deleteLocation(id: number) {
      this.locations = this.locations.filter((location) => location.id !== id);
    },
    setWeatherReports(weatherReport: any) {
      this.weatherReports.push(weatherReport);
    },
    transformLocationData(weatherData: WeatherReportType) {
      return {
        cityName: weatherData.name,
        countryName: weatherData.sys.country,
        weatherDescription: weatherData.weather[0].description,
        feelsLike: weatherData.main.feels_like,
        weatherIcon: getWeatherIconByIconName(weatherData.weather[0].icon),
        temperature: weatherData.main.temp,
        wind: weatherData.wind,
        humidity: weatherData.main.humidity,
        visibility: weatherData.visibility,
        pressure: weatherData.main.pressure,
        clouds: weatherData.clouds.all,
      };
    },
  };
}
export type TStore = ReturnType<typeof createStore>;
