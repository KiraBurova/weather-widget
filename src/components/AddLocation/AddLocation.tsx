import React, { useState } from 'react';
import { observer } from 'mobx-react-lite';

import { TLocation } from '../../types';
import { Icon } from '@iconify/react';
import AddIcon from '@iconify/icons-ic/outline-subdirectory-arrow-left';
import { useStore } from '../../store/context';
import styles from './AddLocation.module.scss';

const AddLocation = observer(() => {
  const store = useStore();
  const { fetchWeatherReportByCityName } = store;

  const [inputtedLocation, setInputtedLocation] = useState('');

  const handleChangeLocation = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value;
    setInputtedLocation(value);
  };

  const handleSubmitLocation = (e: React.FormEvent) => {
    e.preventDefault();
    const location: TLocation = {
      name: inputtedLocation,
      id: Math.random(),
    };
    fetchWeatherReportByCityName(location);
    setInputtedLocation('');
  };

  return (
    <div>
      <span className={styles.formLabel}>Add location:</span>
      <form className={styles.form} onSubmit={handleSubmitLocation}>
        <input autoFocus className={styles.locationInput} value={inputtedLocation} onChange={handleChangeLocation} />
        <button type='submit' className={styles.addButton} disabled={!inputtedLocation}>
          <Icon icon={AddIcon} />
        </button>
      </form>
    </div>
  );
});

export default AddLocation;
