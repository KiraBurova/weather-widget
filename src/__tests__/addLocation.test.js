import React from 'react';

import { render, fireEvent } from '@testing-library/react';

import { StoreProvider } from '../store/context';

import AddLocation from '../components/AddLocation';

describe('<AddLocation />', () => {
  it('change location on input', () => {
    const { container, getByRole } = render(
      <StoreProvider>
        <AddLocation></AddLocation>
      </StoreProvider>,
    );
    const textBox = getByRole('textbox');

    fireEvent.change(textBox, { target: { value: 'Oslo' } });

    expect(textBox.value).toEqual('Oslo');
    expect(container.firstChild).toMatchSnapshot();
  });

  it('submit inputted location', () => {
    const { container, getByTestId, getByRole } = render(
      <StoreProvider>
        <AddLocation></AddLocation>
      </StoreProvider>,
    );
    const textBox = getByRole('textbox');

    fireEvent.change(textBox, { target: { value: 'Oslo' } });

    const form = getByTestId('form');

    fireEvent.submit(form);

    expect(textBox.value).toEqual('');
    expect(container.firstChild).toMatchSnapshot();
  });
});
