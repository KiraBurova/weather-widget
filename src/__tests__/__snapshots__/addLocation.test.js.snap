// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddLocation /> change location on input 1`] = `
<div>
  <span
    class="formLabel"
  >
    Add location:
  </span>
  <form
    class="form"
    data-testid="form"
  >
    <input
      class="locationInput"
      value="Oslo"
    />
    <button
      class="addButton"
      type="submit"
    >
      <svg
        focusable="false"
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style="transform: rotate(360deg);"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 9l1.42 1.42L8.83 14H18V4h2v12H8.83l3.59 3.58L11 21l-6-6l6-6z"
          fill="currentColor"
        />
      </svg>
    </button>
  </form>
</div>
`;

exports[`<AddLocation /> submit inputted location 1`] = `
<div>
  <span
    class="formLabel"
  >
    Add location:
  </span>
  <form
    class="form"
    data-testid="form"
  >
    <input
      class="locationInput"
      value=""
    />
    <button
      class="addButton"
      disabled=""
      type="submit"
    >
      <svg
        focusable="false"
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style="transform: rotate(360deg);"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 9l1.42 1.42L8.83 14H18V4h2v12H8.83l3.59 3.58L11 21l-6-6l6-6z"
          fill="currentColor"
        />
      </svg>
    </button>
  </form>
</div>
`;
